File: clients/client-python/gravitino/client/metadata_object_secret_operations.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.

from typing import Dict

from gravitino.api.metadata_object import MetadataObject
from gravitino.api.secret.supports_secrets import SupportsSecrets
from gravitino.dto.responses.secrets_response import SecretsResponse
from gravitino.exceptions.handlers.credential_error_handler import (
    CREDENTIAL_ERROR_HANDLER,
)
from gravitino.rest.rest_utils import encode_string
from gravitino.utils import HTTPClient


class MetadataObjectSecretOperations(SupportsSecrets):
    _rest_client: HTTPClient
    _request_path: str

    def __init__(
        self,
        metalake_name: str,
        metadata_object: MetadataObject,
        rest_client: HTTPClient,
    ):
        self._rest_client = rest_client
        metadata_object_type = metadata_object.type().value
        metadata_object_fullname = metadata_object.full_name()
        self._request_path = (
            f"api/metalakes/{encode_string(metalake_name)}/objects/{metadata_object_type}/"
            f"{encode_string(metadata_object_fullname)}/secrets"
        )

    def get_secrets(self) -> Dict[str, str]:
        resp = self._rest_client.get(
            self._request_path,
            error_handler=CREDENTIAL_ERROR_HANDLER,
        )
        secret_resp = SecretsResponse.from_json(resp.body, infer_missing=True)
        secret_resp.validate()
        return secret_resp.secrets()

File: clients/client-java/src/main/java/org/apache/gravitino/client/MetadataObjectSecretOperations.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.gravitino.client;

import java.util.Collections;
import java.util.Locale;
import java.util.Map;
import org.apache.gravitino.MetadataObject;
import org.apache.gravitino.dto.responses.SecretsResponse;
import org.apache.gravitino.rest.RESTUtils;
import org.apache.gravitino.secret.SupportsSecrets;

/**
 * The implementation of {@link SupportsSecrets}. This interface will be composited into catalog,
 * schema, and fileset to provide secret property operations for these metadata objects.
 */
class MetadataObjectSecretOperations implements SupportsSecrets {

  private final RESTClient restClient;

  private final String secretsRequestPath;

  MetadataObjectSecretOperations(
      String metalakeName, MetadataObject metadataObject, RESTClient restClient) {
    this.restClient = restClient;
    this.secretsRequestPath =
        String.format(
            "api/metalakes/%s/objects/%s/%s/secrets",
            RESTUtils.encodeString(metalakeName),
            metadataObject.type().name().toLowerCase(Locale.ROOT),
            RESTUtils.encodeString(metadataObject.fullName()));
  }

  @Override
  public Map<String, String> getSecrets() {
    SecretsResponse resp =
        restClient.get(
            secretsRequestPath,
            SecretsResponse.class,
            Collections.emptyMap(),
            ErrorHandlers.credentialErrorHandler());
    resp.validate();
    return resp.getSecrets();
  }
}
